File: World Finals/taking-over-the-world.cpp
// Copyright (c) 2015 kamyu. All rights reserved.

/*
 * Google Code Jam 2015 World Finals - Problem D. Taking Over The World
 * https://code.google.com/codejam/contest/5224486/dashboard#s=p3
 *
 * Time:  O(?)
 * Space: O(?)
 *
 */

#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <map>
#include <algorithm>

using std::cin;
using std::cout;
using std::endl;
using std::string;
using std::vector;
using std::queue;
using std::multimap;
using std::min;

struct e_t {
  int to, cap, rev;
  e_t(int t, int c, int r) : to(t), cap(c), rev(r) {}
};

const int MAX_N = 100;
const int MAX_V = MAX_N * (MAX_N - 1) / 2;
const int GUARD = 1000;

void add_edge(const int i, const int j, const int c,
              vector<vector<e_t>> *adj) {
    (*adj)[i].emplace_back(j, c, (*adj)[j].size());
    (*adj)[j].emplace_back(i, 0, (*adj)[i].size() - 1);
}

bool levelize(const int V, const int S, const int T,
              vector<vector<e_t>> *adj,
              vector<int> *lev) {
    *lev = vector<int>(MAX_V, -1);
    queue<int> que;
    (*lev)[S] = 0;
    que.emplace(S);
    while (!que.empty()) {
        int v = que.front();
        que.pop();
        for (int i = 0; i < (*adj)[v].size(); ++i) {
            const e_t &e = (*adj)[v][i];
            if (e.cap && (*lev)[e.to] == -1) {
                (*lev)[e.to] = (*lev)[v] + 1;
                que.emplace(e.to);
            }
        }
    }
    return (*lev)[T] != -1;
}

int augment(const int S, const int T,
            const int v, const int f,
            const vector<int>& lev,
            vector<vector<e_t>> *adj,
            vector<int> *done) {
    if (v == T || !f) {
        return f;
    }
    for (; (*done)[v] < (*adj)[v].size(); ++(*done)[v]) {
        e_t &e = (*adj)[v][(*done)[v]];
        if (lev[e.to] > lev[v]) {
            int t = augment(S, T, e.to, min(f, e.cap), lev, adj, done);
            if (t > 0) {
                e.cap -= t;
                (*adj)[e.to][e.rev].cap += t;
                return t;
            }
        }
    }
    return 0;
}

int max_flow(const int V, const int S, const int T,
             vector<vector<e_t>> *adj) {
    int f = 0, t;
    vector<int> lev;
    while (levelize(V, S, T, adj, &lev)) {
        vector<int> done(MAX_V);
        while ((t = augment(S, T, S, INT_MAX, lev, adj, &done))) {
            f += t;
        }
    }
    return f;
}

vector<bool> min_cut(const int V, const int S,
                     const vector<vector<e_t>>& adj) {
    vector<bool> vis(V);
    queue<int> que;
    que.emplace(S);
    vis[S] = true;
    while (!que.empty()) {
        int v = que.front();
        que.pop();
        for (const e_t &e : adj[v]) {
            if (e.cap && !vis[e.to]) {
                que.emplace(e.to);
                vis[e.to] = true;
            }
        }
    }
    return vis;
}

vector<int> dijkstra(const vector<bool>& guard,
                     const int N,
                     const vector<vector<int>>& A,
                     const int s) {
    vector<int> dst(N, INT_MAX);
    multimap<int, int> que;
    que.emplace(0, s);
    dst[s] = 0;

    while (!que.empty()) {
        int c = que.begin()->first;
        int v = que.begin()->second;
        que.erase(que.begin());
        if (dst[v] == c) {
            for (int tv = 0; tv < N; ++tv) {
                if (A[v][tv]) {
                    int tc = dst[v] + 1 + (guard[v] ? 1 : 0);
                    if (tc < dst[tv]) {
                        dst[tv] = tc;
                        que.emplace(tc, tv);
                    }
                }
            }
        }
    }
    return dst;
}

int vid(int v, bool o) {
    return v * 2 + (o ? 1 : 0);
}

int taking_over_the_world() {
    int N, M, K;
    cin >> N >> M >> K;

    vector<vector<int>> A(MAX_N, vector<int>(MAX_N));
    for (int i = 0; i < M; ++i) {
        int u, v;
        cin >> u >> v;
        A[u][v] = A[v][u] = true;
    }

    vector<bool> guard(MAX_N);
    while (true) {
        const int V = N * 2;
        const int S = vid(0, false);
        const int T = vid(N - 1, false);

        vector<vector<e_t>> adj(MAX_V);
        for (int v = 0; v < N; ++v) {
            add_edge(vid(v, false), vid(v, true), guard[v] ? GUARD : 1,
                     &adj);
        }

        const vector<int> ds = dijkstra(guard, N, A, 0);
        const vector<int> dt = dijkstra(guard, N, A, N - 1);
        for (int u = 0; u < N; ++u) {
            for (int v = 0; v < N; ++v) {
                if (A[u][v]) {
                    if (ds[u] != -1 && dt[v] != -1) {
                        // Edge (u, v)
                        int td = ds[u] +
                                 (guard[u] ? 1 : 0) + 1 +
                                 (v == N - 1 ? 0 : (guard[v] ? 1 : 0) +
                                 dt[v]);
                        if (td == ds[N - 1]) {
                            add_edge(vid(u, true), vid(v, false), GUARD, &adj);
                        }
                    }
                }
            }
        }

        if (max_flow(V, S, T, &adj) <= K) {
            const vector<bool> mc = min_cut(V, S, adj);
            for (int v = 0; v < N; ++v) {
                if (mc[vid(v, false)] && !mc[vid(v, true)]) {
                    guard[v] = true;
                    --K;
                }
            }
        } else {
            break;
        }
    }
    return dijkstra(guard, N, A, 0)[N - 1];
}

int main() {
    int T;
    cin >> T;

    for (int test = 1; test <= T; ++test) {
        cout << "Case #" << test << ": "
             << taking_over_the_world() << endl;
    }
    return 0;
}

File: World Finals/taking-over-the-world.py
# Copyright (c) 2015 kamyu. All rights reserved.
#
# Google Code Jam 2015 World Finals - Problem D. Taking Over The World
# https://code.google.com/codejam/contest/5224486/dashboard#s=p3
#
# Time:  O(?)
# Space: O(?)
#

from collections import deque
from heapq import heappush, heappop

TO, CAP, REV = 0, 1, 2

def vid(v, o):
    return v * 2 + int(o)


def add_edge(i, j, c, adj):
    adj[i].append([j, c, len(adj[j])])
    adj[j].append([i, 0, len(adj[i]) - 1])


def dijkstra(guard, A, s):
    dst = [float("inf")] * len(A)
    dst[s] = 0
    q = []
    heappush(q, (0, s))
    while q:
        c, v = heappop(q);
        if dst[v] == c:
            for tv in xrange(len(A[v])):
                if A[v][tv]:
                    tc = dst[v] + 1 + int(guard[v]);
                    if tc < dst[tv]:
                        dst[tv] = tc
                        heappush(q, (tc, tv))
    return dst


def levelize(V, S, T, adj, lev):
    for i in xrange(V):
        lev[i] = -1
    lev[S] = 0
    q = deque([S])
    while q:
        v = q.popleft()
        for i in xrange(len(adj[v])):
            to, cap, rev = adj[v][i]
            if cap and lev[to] == -1:
                lev[to] = lev[v] + 1
                q.append(to)

    return lev[T] != -1


def augment(S, T, v, f, lev, adj, done):
    if v == T or not f:
        return f
    while done[v] < len(adj[v]):
        to, cap, rev = adj[v][done[v]]
        if lev[to] > lev[v]:
            t = augment(S, T, to, min(f, cap), lev, adj, done)
            if t > 0:
                adj[v][done[v]][CAP] -= t
                adj[to][rev][CAP] += t
                return t
        done[v] += 1

    return 0


def max_flow(V, S, T, adj):
    f, t = 0, 0
    lev = [-1] * V
    while levelize(V, S, T, adj, lev):
        done = [0] * V
        t = float("inf")
        while t:
            t = augment(S, T, S, float("inf"), lev, adj, done)
            f += t

    return f


def min_cut(V, S, adj):
    vis = [False] * V;
    vis[S] = True;
    q = deque([S]);
    while q:
        v = q.popleft();
        for to, cap, rev in adj[v]:
            if cap and not vis[to]:
                q.append(to)
                vis[to] = True

    return vis


def taking_over_the_world():
    N, M, K = map(int, raw_input().strip().split())
    A = [[0 for _ in xrange(N)] for _ in xrange(N)]
    for i in xrange(M):
        u, v = map(int, raw_input().strip().split())
        A[u][v] = A[v][u] = True

    GUARD = 1000
    guard = [False] * N
    while True:
        V, S, T = N * 2, vid(0, False), vid(N - 1, False)
        adj = [[] for _ in xrange(V)]
        for v in xrange(N):
            add_edge(vid(v, False), vid(v, True), GUARD if guard[v] else 1, adj)

        ds, dt = dijkstra(guard, A, 0), dijkstra(guard, A, N - 1);
        for u in xrange(N):
            for v in xrange(N):
                if A[u][v]:
                    if ds[u] != -1 and dt[v] != -1:
                        # Edge (u, v)
                        td = (ds[u] + int(guard[u])) + \
                             1 + \
                             (0 if v == N - 1 else int(guard[v]) + dt[v])
                        if td == ds[N - 1]:
                            add_edge(vid(u, True), vid(v, False), GUARD, adj)

        if max_flow(V, S, T, adj) <= K:
            mc = min_cut(V, S, adj)
            for v in xrange(N):
                if mc[vid(v, False)] and  not mc[vid(v, True)]:
                    guard[v] = True
                    K -= 1
        else:
            break

    return dijkstra(guard, A, 0)[N - 1]


for case in xrange(input()):
    print "Case #%d: %d" % (case+1, taking_over_the_world())
